### PR TITLE
Rup set diags

### PR DESCRIPTION
**Pull request by chrisbc**
not clean :( but we tried !

### CODE REVIEW

File: src/org/opensha/sha/earthquake/faultSysSolution/ruptures/util/RupSetDiagnosticsPageGen.java [hunk 1]
```diff
@@ -258,53 +258,53 @@ public static void main(String[] args) throws IOException, DocumentException {
...
 	
-	private FaultSystemRupSet inputRupSet;
-	private FaultSystemSolution inputSol;
-	private String inputName;
-	private PlausibilityConfiguration inputConfig;
-	private List<ClusterRupture> inputRups;
-	private RuptureConnectionSearch inputSearch;
-	private HashSet<UniqueRupture> inputUniques;
-	private ClusterConnectionStrategy inputConnStrat;
+	public FaultSystemRupSet inputRupSet;
```
Comment (voj) [inline src/org/opensha/sha/earthquake/faultSysSolution/ruptures/util/RupSetDiagnosticsPageGen.java]: this probably shouldn't be public

File: src/org/opensha/sha/earthquake/faultSysSolution/ruptures/util/RupSetDiagnosticsPageGen.java [hunk 2]
```diff
@@ -685,8 +685,14 @@ public void generatePage() throws IOException {
 					summary.compMeta.scalars.add(new ScalarRange(compScalars));
 					compScalarVals.add(compScalars);
 				}
-				plotRuptureHistograms(resourcesDir, "hist_"+scalar.name(), table, inputScalars,
-						inputUniques, compScalars, compUniques);
+				try {
+					plotRuptureHistograms(resourcesDir, "hist_"+scalar.name(), table, inputScalars,
+							inputUniques, compScalars, compUniques);
+				} catch( IllegalStateException wee) {
```
Comment (voj) [inline src/org/opensha/sha/earthquake/faultSysSolution/ruptures/util/RupSetDiagnosticsPageGen.java]: if you want to leave this in, the block should be formatted, and I would recommend printing out the exception too.

File: src/org/opensha/sha/earthquake/faultSysSolution/ruptures/util/RupSetDiagnosticsPageGen.java [hunk 3]
```diff
@@ -943,30 +949,37 @@ else if (f == 1d)
...
+						plotConnectivityHistogram(resourcesDir, "sect_connectivity_hist_rates_comp",
+								getTruncatedTitle(compName)+" Connectivity", compJumps, compUniqueJumps, maxConnDist,
+								COMP_COLOR, true, false);
+						plotConnectivityHistogram(resourcesDir, "sect_connectivity_hist_rates_comp_log",
+								getTruncatedTitle(compName)+" Connectivity", compJumps, compUniqueJumps, maxConnDist,
+								COMP_COLOR, true, true);
+					}
 				}
-			}
+			} catch( IllegalStateException ouch) {
```
Comment (voj) [inline src/org/opensha/sha/earthquake/faultSysSolution/ruptures/util/RupSetDiagnosticsPageGen.java]: same as above

File: src/org/opensha/sha/earthquake/faultSysSolution/ruptures/util/RupSetDiagnosticsPageGen.java [hunk 4]
```diff
@@ -1303,32 +1343,10 @@ else if (minMag >= 6.5)
...
-						table.addColumn("![Scatter](resources/"+linearPlot.getName()+")");
-						table.addColumn("![Scatter](resources/"+logPlot.getName()+")");
-						table.finalizeLine();
-					}
-				}
-				lines.addAll(table.build());
-				lines.add("");
-			}
-		}
+		} catch( Exception ouch) {
```
Comment (voj) [inline src/org/opensha/sha/earthquake/faultSysSolution/ruptures/util/RupSetDiagnosticsPageGen.java]: ditto

File: src/scratch/UCERF3/FaultSystemRupSet.java [hunk 5]
```diff
@@ -113,7 +113,7 @@ protected FaultSystemRupSet() {
 	 * Initialize from another rupSet
 	 * @param rupSet
 	 */
-	protected void init(FaultSystemRupSet rupSet) {
+	public void init(FaultSystemRupSet rupSet) {
```
Comment (voj) [inline src/scratch/UCERF3/FaultSystemRupSet.java]: could this be protected?

File: src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java [hunk 6]
```diff
@@ -252,12 +252,69 @@ public double getArea(double mag, double origWidth) {
 			return map;
 		}
 
-	};
+	},
+	
+	TMG_SUB_2017("Thingbaijam et al.(2017) Subduction", "TMG_SUB_2017") {
+		// units of the input dimensions are in m or m^2 
+		public double getAveSlip(double area, double length, double origWidth) {
+			tmg_sub_magArea.setRake(90.0d); // interface
```
Comment (voj) [inline src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java:252]: I had a look at how other scalingrelationships handle this, and none of them store state like this. In fact it seems they make sure they're threadsafe. I think I did accept this in my review for `TMG2017SubMagAreaRel`, but now I think that was a mistake. `TMG2017SubMagAreaRel` should not have state.

File: src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java [hunk 7]
```diff
@@ -252,12 +252,69 @@ public double getArea(double mag, double origWidth) {
...
+			return tmg_sub_magArea.getMedianArea(mag)*1e6;
+		}
+		
+		
+		@Override
+		public double getRelativeWeight(InversionModels im) {
+			// weight needs to be updated
+			return 1.0d; 
+
```
Comment (voj) [inline src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java:272]: there are some stray newlines in this file

File: src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java [hunk 8]
```diff
@@ -252,12 +252,69 @@ public double getArea(double mag, double origWidth) {
...
+		}
+	},
+	
+	TMG_CRU_2017("Thingbaijam et al.(2017) Crustal", "TMG_CRU_2017") {
+		// units of the input dimensions are in m or m^2 
+		public double getAveSlip(double area, double length, double origWidth) {
+			tmg_cru_magArea.setRake(0.0d); //strike-slip faulting
```
Comment (voj) [inline src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java]: same comment re statelessness

File: src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java [hunk 9]
```diff
@@ -273,6 +330,7 @@ private ScalingRelationships(String name, String shortName) {
 	 * @param length (m)
 	  * @param origWidth (m) - the original down-dip width (before reducing by aseismicity factor)
 	 * @return
+	 * @throws Exception
```
Comment (voj) [inline src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java]: it doesn't throw?

File: src/scratch/UCERF3/griddedSeismicity/FaultPolyMgr.java [hunk 10]
```diff
@@ -70,6 +69,11 @@ public GriddedRegion getRegion() {
 		return region;
 	}
 	
+	private FaultPolyMgr setRegion(GriddedRegion region) {
```
Comment (voj) [inline src/scratch/UCERF3/griddedSeismicity/FaultPolyMgr.java]: didn't we delete this?

**Review by voj — CHANGES_REQUESTED** [review]
I think we have a problem with the statefulness of the scalingrelationships :/

File: src/org/opensha/sha/earthquake/faultSysSolution/ruptures/util/RupSetDiagnosticsPageGen.java (hunk 2 above)
Comment (chrisbc) [reply src/org/opensha/sha/earthquake/faultSysSolution/ruptures/util/RupSetDiagnosticsPageGen.java]: I've been nursing this file along, and committed it accidentally. I'll fix up the exception handlers as you suggest :)

File: src/scratch/UCERF3/FaultSystemRupSet.java (hunk 5 above)
Comment (chrisbc) [reply src/scratch/UCERF3/FaultSystemRupSet.java]: yes

File: src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java (hunk 6 above)
Comment (chrisbc) [reply src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java:252]: Yes, I see what you're saying. I think the problem stems from the base class...

```
package org.opensha.commons.calc.magScalingRelations;
....
public abstract class MagScalingRelationship 
```

as this is where the setRake() method is defined. Hard to know what the impact would be of changing this API. A couple of suggestions...

 - make the setRake method protected or even private.
 - create a constructor that passes in rake.

all that said, the way it's being used now for TMG shouldn't cause problems, thought that's more good luck than anything :)

Comment (chrisbc) [reply src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java:252]: I suggest we raise a separate ticket for this one.

File: src/scratch/UCERF3/griddedSeismicity/FaultPolyMgr.java (hunk 10 above)
Comment (chrisbc) [reply src/scratch/UCERF3/griddedSeismicity/FaultPolyMgr.java]: No but we should have ... I'll do it

File: src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java (hunk 7 above)
Comment (chrisbc) [reply src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java:272]: fixed

File: src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java (hunk 9 above)
Comment (chrisbc) [reply src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java]: Done

File: src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java (hunk 6 above)
Comment (chrisbc) [reply src/scratch/UCERF3/enumTreeBranches/ScalingRelationships.java:252]: here's the ticket https://github.com/GNS-Science/opensha-ucerf3/issues/25

**Pull request merged**